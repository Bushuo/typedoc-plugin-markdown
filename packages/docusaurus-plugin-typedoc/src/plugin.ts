import * as path from 'path';

import { LoadContext } from '@docusaurus/types';
import { Application, TSConfigReader, TypeDocReader } from 'typedoc';

import { DocusaurusFrontMatterComponent } from './front-matter';
import { writeSidebar } from './sidebar';
import DocusaurusTheme from './theme/theme';
import { PluginOptions } from './types';

const DEFAULT_PLUGIN_OPTIONS: PluginOptions = {
  docsRoot: 'docs',
  out: 'api',
  sidebar: {
    fullNames: false,
    sidebarFile: 'typedoc-sidebar.js',
    globalsLabel: 'Index',
    readmeLabel: 'Readme',
  },
  globalsTitle: undefined,
  readmeTitle: undefined,
};

let app: Application;

export default async function pluginDocusaurus(
  context: LoadContext,
  opts: Partial<PluginOptions>,
) {
  const { siteDir } = context;

  /**
   * Configure options
   */
  const options = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...opts,
    ...(opts.sidebar && {
      sidebar: {
        ...DEFAULT_PLUGIN_OPTIONS.sidebar,
        ...opts.sidebar,
      },
    }),
  };

  /**
   * Initialize and build app
   * - currently we can't re-compile with devServer due to infinate render loop
   * - @TODO look if configureWebpack lifecycle can be updated to handle more gracefully
   */
  if (!app) {
    app = new Application();

    // TypeDoc options
    const typedocOptions = Object.keys(options).reduce((option, key) => {
      if (![...['id'], ...Object.keys(DEFAULT_PLUGIN_OPTIONS)].includes(key)) {
        option[key] = options[key];
      }
      return option;
    }, {});

    app.options.addReader(new TypeDocReader());
    app.options.addReader(new TSConfigReader());

    // bootstrap TypeDoc app
    app.bootstrap({
      // filtered TypeDoc options
      ...typedocOptions,
      // TypeDoc plugins
      plugin: [
        ...['typedoc-plugin-markdown'],
        ...(opts.plugin
          ? opts.plugin.filter((name) => name !== 'typedoc-plugin-markdown')
          : []),
      ],
      // add docusaurus theme
      theme: path.resolve(__dirname, 'theme'),
    });

    // add frontmatter component
    app.renderer.addComponent(
      'docusaurus-frontmatter',
      new DocusaurusFrontMatterComponent(app.renderer, options),
    );

    // return the generated reflections
    const project = app.convert();

    // if project is undefined typedoc has a problem - error logging will be supplied by typedoc.
    if (!project) {
      return;
    }

    // construct outputDirectory path
    const outputDirectory = path.resolve(
      siteDir,
      options.docsRoot,
      options.out,
    );

    // generate the static docs
    await app.generateDocs(project, outputDirectory);

    // write the sidebar (if applicable)
    if (options.sidebar) {
      const theme = app.renderer.getComponent('theme') as DocusaurusTheme;
      writeSidebar(
        options.disableOutputCheck || theme.isOutputDirectory(outputDirectory),
        siteDir,
        options.out,
        options.sidebar,
        theme.getNavigation(project),
      );
    }
  }

  // we need to generate the sidebar before any available lifecycle apis
  return {
    name: 'docusaurus-plugin-typedoc',
  };
}
