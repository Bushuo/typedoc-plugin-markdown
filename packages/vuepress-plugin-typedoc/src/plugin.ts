import * as path from 'path';

import { Application, ProjectReflection } from 'typedoc';
import MarkdownPlugin from 'typedoc-plugin-markdown';

import { FrontMatterComponent } from './front-matter';
import { addOptions, getOptions } from './options';
import { render } from './render';
import { getSidebarJson } from './sidebar';
import { PluginOptions } from './types';

let app: Application;
let project: ProjectReflection | undefined;

export const typedocPlugin = (opts: PluginOptions, ctx: any) => {
  const options = getOptions(opts);

  app = new Application();

  MarkdownPlugin(app);

  app.renderer.render = render;

  addOptions(app);

  app.bootstrap({ ...options, theme: path.resolve(__dirname) });

  app.renderer.addComponent('fm', new FrontMatterComponent(app.renderer));

  project = app.convert();

  // if project is undefined typedoc has a problem - error logging will be supplied by typedoc.
  if (!project) {
    return;
  }

  const outputDirectory = ctx.sourceDir + '/' + options.out;
  app.generateDocs(project, outputDirectory);

  if (options.watch) {
    app.convertAndWatch(async (project) => {
      app.generateDocs(project, outputDirectory);
    });
  }

  return {
    name: 'vuepress-plugin-typedoc',

    async enhanceAppFiles() {
      if (!options.sidebar) {
        return;
      }
      const theme = app.renderer.theme as any;
      const navigation = theme.getNavigation(project);
      const sidebarJson = JSON.stringify({
        [`/${options.out}/`]: getSidebarJson(navigation, options),
      });
      return {
        name: 'typedoc-sidebar',
        content: `export default ({ siteData, options }) => {
          siteData.themeConfig.sidebarDepth = 0;
          siteData.themeConfig.sidebar = Object.assign({},siteData.themeConfig.sidebar,${sidebarJson});
        }`,
      };
    },
  };
};
