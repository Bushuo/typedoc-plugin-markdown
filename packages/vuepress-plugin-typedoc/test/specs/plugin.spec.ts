import * as fs from 'fs-extra';
import * as tmp from 'tmp';

import { typedocPlugin } from '../../dist/plugin';

tmp.setGracefulCleanup();

function bootstrap(tmpobj: tmp.DirResult, customOptions = {}) {
  const options = {
    id: 'default',
    logger: 'none',
    entryPoints: ['../typedoc-plugin-markdown/test/stubs/src/theme.ts'],
    tsconfig: '../typedoc-plugin-markdown/test/stubs/tsconfig.json',
  } as any;

  const plugin = typedocPlugin(
    { ...options, ...customOptions },
    { sourceDir: tmpobj.name },
  );
  return plugin;
}

describe(`(render)`, () => {
  let tmpobj;
  beforeAll(async () => {
    tmpobj = tmp.dirSync();
    bootstrap(tmpobj);
  });
  test(`should write docs`, async () => {
    const files = fs.readdirSync(tmpobj.name + '/api');
    expect(files).toMatchSnapshot();
  });
  test(`should write doc`, () => {
    const sidebar = fs.readFileSync(tmpobj.name + '/api/README.md');
    expect(sidebar.toString()).toMatchSnapshot();
  });
});

describe(`(sidebars)`, () => {
  let tmpobj;
  beforeAll(async () => {
    tmpobj = tmp.dirSync();
    bootstrap(tmpobj);
  });
  test(`should generate default`, async () => {
    const plugin = bootstrap(tmpobj, { sidebar: { parentCategory: 'none' } });
    const enhancedFiles = await plugin.enhanceAppFiles();
    expect(enhancedFiles).toMatchSnapshot();
  });

  test(`should generate with parent category and fullNames`, async () => {
    const plugin = bootstrap(tmpobj, {
      sidebar: { fullNames: true },
    });
    const enhancedFiles = await plugin.enhanceAppFiles();
    expect(enhancedFiles).toMatchSnapshot();
  });

  test(`should skip sidebar`, async () => {
    const plugin = bootstrap(tmpobj, { sidebar: null });
    const enhancedFiles = await plugin.enhanceAppFiles();
    expect(enhancedFiles).toBeUndefined();
  });
});
