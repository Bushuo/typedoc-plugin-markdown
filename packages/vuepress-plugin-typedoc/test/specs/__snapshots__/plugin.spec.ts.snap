// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(render) should write doc 1`] = `
"---
title: \\"typedoc-plugin-markdown\\"
---

# MY README

> Some block quote

<!-- some comments -->

Some text
"
`;

exports[`(render) should write docs 1`] = `
Array [
  "README.md",
  "classes",
  "enums",
  "interfaces",
  "modules",
  "modules.md",
]
`;

exports[`(sidebars) should generate default 1`] = `
Object {
  "content": "export default ({ siteData, options }) => {
          siteData.themeConfig.sidebarDepth = 0;
          siteData.themeConfig.sidebar = Object.assign({},siteData.themeConfig.sidebar,{\\"/api/\\":[[\\"/api/\\",\\"Readme\\"],[\\"modules\\",\\"Exports\\"],{\\"title\\":\\"Namespaces\\",\\"children\\":[[\\"modules/namespacea\\",\\"namespaceA\\"],[\\"modules/namespaceb\\",\\"namespaceB\\"]]},{\\"title\\":\\"Enumerations\\",\\"children\\":[[\\"enums/enumitema\\",\\"enumItemA\\"],[\\"enums/enumitemb\\",\\"enumItemB\\"]]},{\\"title\\":\\"Classes\\",\\"children\\":[[\\"classes/namespacea.somenestedclass\\",\\"SomeNestedClass\\"],[\\"classes/classitema\\",\\"ClassItemA\\"],[\\"classes/classitemb\\",\\"ClassItemB\\"]]},{\\"title\\":\\"Interfaces\\",\\"children\\":[[\\"interfaces/interfaceitema\\",\\"InterfaceItemA\\"],[\\"interfaces/interfaceitemb\\",\\"InterfaceItemB\\"]]}]});
        }",
  "name": "typedoc-sidebar",
}
`;

exports[`(sidebars) should generate with parent category and fullNames 1`] = `
Object {
  "content": "export default ({ siteData, options }) => {
          siteData.themeConfig.sidebarDepth = 0;
          siteData.themeConfig.sidebar = Object.assign({},siteData.themeConfig.sidebar,{\\"/api/\\":[{\\"title\\":\\"Parent Category\\",\\"children\\":[[\\"/api/\\",\\"Readme\\"],[\\"modules\\",\\"Exports\\"],{\\"title\\":\\"Namespaces\\",\\"children\\":[[\\"modules/namespacea\\",\\"namespaceA\\"],[\\"modules/namespaceb\\",\\"namespaceB\\"]]},{\\"title\\":\\"Enumerations\\",\\"children\\":[[\\"enums/enumitema\\",\\"enumItemA\\"],[\\"enums/enumitemb\\",\\"enumItemB\\"]]},{\\"title\\":\\"Classes\\",\\"children\\":[[\\"classes/namespacea.somenestedclass\\",\\"namespaceA.SomeNestedClass\\"],[\\"classes/classitema\\",\\"ClassItemA\\"],[\\"classes/classitemb\\",\\"ClassItemB\\"]]},{\\"title\\":\\"Interfaces\\",\\"children\\":[[\\"interfaces/interfaceitema\\",\\"InterfaceItemA\\"],[\\"interfaces/interfaceitemb\\",\\"InterfaceItemB\\"]]}]}]});
        }",
  "name": "typedoc-sidebar",
}
`;
